wave_utils/include/wave/utils/pose_cov_comp.hpp

```diff
@@ -0,0 +1,233 @@
+    Vector3 getPosition() const;
+    Vector3 getYPR() const;
+    Vector4 getQuaternion() const;
+    Vector7 getPoseQuaternion() const;
+    Eigen::Affine3d getTransformMatrix() const;
+};
+
+/** Calculates the pose composition of two poses and the estimated covariance.
+ *  Based on [1].
+ *
+ *  @param p1 first pose with covariance
+ *  @param p2 second pose with covariance
+ *
+ *  @return composed pose with predicted covariance
+ *
+ */
+PoseWithCovariance composePose(PoseWithCovariance &p1, PoseWithCovariance &p);
+
+/** The Jacobian of quaternion normalization function. Quaternion in the form of
+ *  [qr, qx, qt, qz]
+ *  Equation (1.7)
+ *
+ *  @param q quaternion vector
+ *
+ *  @return quaternion normal wrt q Jacobian
+ */
+Matrix4x4 jacobian_Quat_Norm_wrt_q(const Vector4 &q);
+
+/** The Jacobian of normalized quaternion to rpy function.
+ *  Equation (2.9) to Equation (2.10)
+ *
+ *  @param q quaternion vector
+ *
+ *  @return quaternion normal to roll pitch yaw wrt q Jacobian
+ */
+Matrix3x4 jacobian_Quat_Norm_to_Rpy_wrt_q(const Vector4 &q);
+
+/** The Jacobian of p7 to p6 conversion.
+ *  Equation (2.12)
+ *
+ *  @param p pose with position and quaternion
+ *
+ *  @return p7 to p6 wrt p Jacobian
+ */
+Matrix6x7 jacobian_p7_to_p6_wrt_p(const Vector7 &p);
+
+/** Jacobian of composing a point to a p7.
+ *  Equation (3.8)
+ *
+ *  @param p pose with position and quaternion
+ *  @param a position
+ *  @return p7 point composition wrt p Jacobian
+ */
+Matrix3x7 jacobian_p7_Point_Composition_wrt_p(const Vector7 &p,
+                                              const Vector3 &a);
+/** Jacobian of the composition of p7 poses.
+ *  Equation (5.8)
+ *
+ *  @param p1 pose with position and quaternion
+ *  @param p2 pose with position and quaternion
+ *  @return p7 p7 composition wrt p1 Jacobian
+ */
+Matrix7x7 jacobian_p7_p7_Composition_wrt_p1(const Vector7 &p1,
+                                            const Vector7 &p2);
+
+/** Jacobian of composing a point to a p7.
+ *  Equation (3.10)
+ *
+ *  @param p pose with position and quaternion
+ *  @param a position
+ *  @return p7 point composition wrt a Jacobian
+ */
+Matrix3x3 jacobian_p7_Point_Composition_wrt_a(const Vector7 &p,
+                                              const Vector3 &a);
+
+/** Jacobian of the composition of p7 poses.
+ *  Equation (5.9)
+ *
+ *  @param p1 pose with position and quaternion
+ *  @param p2 pose with position and quaternion
+ *
+ *  @return p7 p7 composition wrt p2 Jacobian
+ */
+Matrix7x7 jacobian_p7_p7_Composition_wrt_p2(const Vector7 &p1,
+                                            const Vector7 &p2);
+
+/** Jacobian of converting a p6 to a p7.
+ *  Equation (2.8)
+ *
+ *  @param p pose with position and ypr
+ *  @return p6 to p7 wrt p Jacobian
+ */
+Matrix7x6 jacobian_p6_to_p7_wrt_p(const Vector6 &p);
+}
+
+/** NOTE: wave_geometry/ has a similar implementation of the following functions
```

Comment:
Doxygen will try to apply this documentation to the whole `wave` namespace

Reply:
So would it work if  I put the comment above the `pose_comp` namespace?

Reply:
Moved to go under nested namespace `pose_comp`